File: lib/refusal.ts
/**
 * The refusal sentence and its detector, in a module with NO dependencies so both the
 * server (chat route, query log, eval harness) and the CLIENT bundle can import it.
 *
 * lib/retrieve.ts pulls in Supabase, OpenAI and Cohere — importing it from a "use client"
 * component would drag all of that into the browser bundle. This split exists for that
 * reason, not for tidiness.
 *
 * One definition, used everywhere: if these ever diverged, the UI, the production log and
 * the eval harness would disagree about whether the same answer was a refusal.
 */
export const REFUSAL_MESSAGE = "I don't have information about that in the documentation. I can help with AI SDK docs. Ask me about those and I'll help.";

/**
 * Does this answer REFUSE — not merely contain a refusal somewhere.
 *
 * The first version used includes(), which called a partial answer a refusal:
 *     "To stream text, use streamText... I don't have information about the second
 *      part of your question in the documentation."
 * ...was scored as a full refusal. One wrong predicate broke three things at once —
 * the eval marked correct behaviour FAIL, the UI printed "I couldn't answer this"
 * under an answer that did, and the production query log recorded partial answers as
 * refusals, poisoning the suspicious_refusals view that exists to find real ones.
 *
 * The prompt instructs the model to LEAD with the sentence when it cannot answer, so
 * position carries the meaning: a refusal starts with it. A refusal followed by a
 * polite offer still counts; an answer that declines one clause at the end does not.
 *
 * The marker is the CORE sentence — "I don't have information about that in the
 * documentation" — not the whole message. The polite tail ("I can help with AI SDK docs…") is
 * an offer the model drops or garbles at will, and requiring it made a must-refuse case flaky:
 * the model refused every time, but one run in three said it without the tail and the exact
 * 100-character match missed. Matching the core keeps the meaning and drops the noise. It is
 * still positional (startsWith / "the documentation" prefix), so a partial answer that answers
 * first and declines a clause at the end still does not count — the Day-12 includes() bug
 * cannot come back through here.
 *
 * For a specific out-of-scope topic the model names what it can't cover. Observed shapes:
 *   a) "The documentation doesn't cover fine-tuning. <core refusal>[tail]"
 *   b) "The documentation doesn't cover fine-tuning a model with the Vercel AI SDK."
 *   c) "The documentation doesn't cover fine-tuning …. I can help with AI SDK docs. Ask me
 *      about those and I'll help."  — prefix + polite tail, NO core sentence.
 * The first detector listed (a) and (b) one by one, and then (c) showed up (1 run in 3 on
 * guard-finetune) and slipped through. Enumerating shapes is a losing game, so the rule is now
 * COMPOSITIONAL: a reply is a refusal when its first sentence is a negative statement about the
 * docs ("The documentation doesn't cover/mention/… X") and EVERY sentence after it is one of the
 * canonical refusal sentences from REFUSAL_MESSAGE — or there are none. That covers (a), (b),
 * (c) and any recombination of them, and nothing else: a leak ("…doesn't cover X. The capital
 * of France is Paris.") has a non-canonical sentence and fails; the topic clause may not carry
 * clause punctuation (",;:") so "…doesn't cover X, but here's how: …" cannot hide inside it.
 */

/** The load-bearing first sentence of REFUSAL_MESSAGE. Derived, so the two can't drift. */
const REFUSAL_CORE = REFUSAL_MESSAGE.split(". ")[0];

const norm = (s: string) => s.toLowerCase().replace(/[^a-z0-9 ]/g, "").replace(/\s+/g, " ").trim();

/** Sentence boundary: terminal punctuation followed by whitespace. norm() strips the punctuation. */
const SENTENCE_BOUNDARY = /[.!?]\s+/;

/** Every sentence of REFUSAL_MESSAGE, normalised. Derived — reword the message and this follows. */
const CANONICAL_SENTENCES = new Set(REFUSAL_MESSAGE.split(SENTENCE_BOUNDARY).map(norm).filter(Boolean));

/**
 * "The documentation doesn't cover <topic>." as an opening sentence. The verb list is closed on
 * purpose — "does not require…" or "does not throw…" can open a real answer; these verbs cannot.
 * The topic may not contain ",;:" (see above). The final period is optional only at the very
 * end of the reply, so "…doesn't cover X" alone still counts.
 */
const NEGATIVE_OPENER =
    /^the documentation (?:does(?:n't|n’t| not) (?:cover|mention|include|discuss|describe|contain|provide|address|explain)|has no (?:information|details?|guidance))\b[^.,;:]*(?:[.!]\s*|$)/i;

export function isRefusal(answer: string): boolean {
    const raw = answer.trim();
    const n = norm(raw);
    const core = norm(REFUSAL_CORE);
    // The canonical refusal, or a "The documentation…" opener followed by the core sentence
    // anywhere. The second form is the one place includes() is safe: the opener already rules
    // out an answer-first reply, and the core sentence contradicts any answer before it.
    if (n.startsWith(core) || (n.startsWith("the documentation") && n.includes(core))) return true;

    // Compositional: negative opener + nothing but canonical refusal sentences after it.
    const opener = NEGATIVE_OPENER.exec(raw);
    if (!opener) return false;
    const rest = raw.slice(opener[0].length).trim();
    if (rest === "") return true;
    return rest.split(SENTENCE_BOUNDARY).map(norm).filter(Boolean).every((s) => CANONICAL_SENTENCES.has(s));
}

File: agent/src/copilot_agent/refusal.py
"""The refusal sentence and its detector: the Python port of lib/refusal.ts.

The query log (query_log.py) records, per request, whether the answer REFUSED. The eval set is
mined from that column (the suspicious_refusals view, db/002_query_log.sql), so the service must
call an answer a refusal exactly when the TypeScript route would. is_refusal() is a port of
isRefusal() and tests/test_refusal.py compares the two verdict by verdict, through a golden file
written by the TypeScript function itself (scripts/experiments/refusal-verdicts.ts).

lib/refusal.ts explains the rule (positional, then compositional) and the bugs that shaped it.
What this file adds is the porting traps, each one a place where the same regex text means
something else in Python:

  - JavaScript's `$` without the m flag is the end of the input; Python's `$` also matches just
    before a trailing newline. Here: `\\Z`, the literal translation. (No test can tell the two
    apart in this function: the answer is trimmed before the pattern sees it. A mutation run
    showed exactly that, so the golden has no case claiming otherwise.)
  - JavaScript's `\\s` and trim() use one whitespace set; Python's `\\s` and strip() use another
    (U+FEFF is whitespace only in JavaScript, \\x1c to \\x1f only in Python). Here: the JavaScript
    set, spelled out (planner.JS_WHITESPACE).
  - A JavaScript regex without the u flag has ASCII `\\b`, and its /i does not fold a non-ASCII
    letter into an ASCII one. Python's str patterns are Unicode by default: `\\b` sees accented
    letters as word characters, and re.IGNORECASE matches the long s as "s" (and the Kelvin
    sign as "k"). Here: re.ASCII.
"""

import re

from copilot_agent.planner import JS_WHITESPACE, js_trim

# lib/refusal.ts, verbatim; tests/test_ts_parity.py pins it. The prompt quotes it
# (generation.py), and the canned off-topic reply is it (graph.py).
REFUSAL_MESSAGE = (
    "I don't have information about that in the documentation. "
    "I can help with AI SDK docs. Ask me about those and I'll help."
)

# JavaScript's \s, as a character class body.
_WS = re.escape(JS_WHITESPACE)

# The load-bearing first sentence of REFUSAL_MESSAGE. Derived, as in TypeScript.
REFUSAL_CORE = REFUSAL_MESSAGE.split(". ")[0]

# Terminal punctuation followed by whitespace.
_SENTENCE_BOUNDARY = re.compile(rf"[.!?][{_WS}]+")

# "The documentation doesn't cover <topic>." as an opening sentence (lib/refusal.ts has why the
# verb list is closed and why the topic may not hold ",;:"). Used with .match(), which anchors
# at the start like the JavaScript ^.
_NEGATIVE_OPENER = re.compile(
    r"the documentation (?:does(?:n't|n\u2019t| not) "
    r"(?:cover|mention|include|discuss|describe|contain|provide|address|explain)"
    r"|has no (?:information|details?|guidance))"
    rf"\b[^.,;:]*(?:[.!][{_WS}]*|\Z)",
    re.IGNORECASE | re.ASCII,
)


def _norm(text: str) -> str:
    # Lowercase, keep only [a-z0-9 ], collapse spaces. After the second step only spaces are
    # left to collapse, so JavaScript's \s+ and trim() act on spaces alone here.
    kept = re.sub(r"[^a-z0-9 ]", "", text.lower())
    return re.sub(" +", " ", kept).strip(" ")


_CANONICAL_SENTENCES = {
    s for s in (_norm(part) for part in _SENTENCE_BOUNDARY.split(REFUSAL_MESSAGE)) if s
}


def is_refusal(answer: str) -> bool:
    """Does this answer REFUSE, not merely contain a refusal somewhere (lib/refusal.ts)."""
    raw = js_trim(answer)
    n = _norm(raw)
    core = _norm(REFUSAL_CORE)
    if n.startswith(core) or (n.startswith("the documentation") and core in n):
        return True

    opener = _NEGATIVE_OPENER.match(raw)
    if opener is None:
        return False
    rest = js_trim(raw[opener.end() :])
    if rest == "":
        return True
    sentences = (_norm(s) for s in _SENTENCE_BOUNDARY.split(rest))
    return all(s in _CANONICAL_SENTENCES for s in sentences if s)
